br/pkg/checkpoint/storage.go

```diff
@@ -0,0 +1,333 @@
+	if err != nil {
+		return checkpointChecksum, pastDureTime, errors.Trace(err)
+	}
+	for _, content := range checkpointChecksums {
+		if err := parseCheckpointChecksum(content, checkpointChecksum, &pastDureTime); err != nil {
+			return checkpointChecksum, pastDureTime, errors.Trace(err)
+		}
+	}
+	return checkpointChecksum, pastDureTime, nil
+}
+
+func initCheckpointTable(ctx context.Context, se glue.Session, dbName string, checkpointTableNames []string) error {
+	if err := se.ExecuteInternal(ctx, "CREATE DATABASE IF NOT EXISTS %n;", dbName); err != nil {
+		return errors.Trace(err)
+	}
+	for _, tableName := range checkpointTableNames {
+		if err := se.ExecuteInternal(ctx, createCheckpointTable, dbName, tableName); err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func insertCheckpointMeta[T any](ctx context.Context, se glue.Session, dbName string, tableName string, meta *T) error {
+	data, err := json.Marshal(meta)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if err := se.ExecuteInternal(ctx, createCheckpointMetaTable, dbName, tableName); err != nil {
+		return errors.Trace(err)
+	}
+	err = chunkInsertCheckpointData(data, func(segmentId uint64, chunk []byte) error {
+		err := se.ExecuteInternal(ctx, insertCheckpointMetaSQLTemplate, dbName, tableName, segmentId, chunk)
+		return errors.Trace(err)
+	})
+	return errors.Trace(err)
+}
+
+func selectCheckpointMeta[T any](
+	ctx context.Context,
+	execCtx sqlexec.RestrictedSQLExecutor,
+	dbName string, tableName string,
+	meta *T,
+) error {
+	rows, _, errSQL := execCtx.ExecRestrictedSQL(
+		kv.WithInternalSourceType(ctx, kv.InternalTxnBR),
+		nil,
+		selectCheckpointMetaSQLTemplate,
+		dbName, tableName,
+	)
+	if errSQL != nil {
+		return errors.Annotatef(errSQL, "failed to get checkpoint metadata from table %s.%s", dbName, tableName)
+	}
+	if len(rows) == 0 {
+		return errors.Errorf(
+			"get the empty checkpoint meta, the checkpoint is incomplete from table %s.%s", dbName, tableName)
+	}
+
+	data := make([]byte, 0, len(rows)*CheckpointIdMapBlockSize)
+	for i, row := range rows {
+		segmentId, chunk := row.GetUint64(0), row.GetBytes(1)
+		if uint64(i) != segmentId {
+			return errors.Errorf(
+				"the checkpoint metadata is incomplete from table %s.%s at segment %d", dbName, tableName, segmentId)
+		}
+		data = append(data, chunk...)
+	}
+	err := json.Unmarshal(data, meta)
+	return errors.Trace(err)
+}
+
+func dropCheckpointTables(
+	ctx context.Context,
+	dom *domain.Domain,
+	se glue.Session,
+	dbName string, tableNames []string,
+) error {
+	for _, tableName := range tableNames {
+		if err := se.ExecuteInternal(ctx, "DROP TABLE IF EXISTS %n.%n;", dbName, tableName); err != nil {
+			return errors.Trace(err)
+		}
+	}
+	// check if any user table is created in the checkpoint database
+	tables, err := dom.InfoSchema().SchemaTableInfos(ctx, model.NewCIStr(dbName))
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if len(tables) > 0 {
```

Comment:
Another thing we can do later is stop restore these _TiDB_xxx tables for any restoration. user might take a backup with half restoration?